dependency-specification.rst

```diff
@@ -0,0 +1,376 @@
+--------
+
+A dependency specification always specifies a distribution name. It may
+include extras, which expand the dependencies of the named distribution to
+enable optional features. The version installed can be controlled using
+version limits, or giving the URL to a specific artifact to install. Finally
+the dependency can be made conditional using environment markers.
+
+Grammar
+-------
+
+We first cover the grammar briefly and then drill into the semantics of each
+section later.
+
+A distribution specification is written in ASCII text. We use ABNF [#abnf]_ to
+provide a precise grammar. Specifications may have comments starting with a
+'#' and running to the end of the line::
+
+    comment       = "#" *(WSP / VCHAR)
+
+Specifications may be spread across multiple lines if desired using
+continuations - a single backslash followed by a new line ('\\\\n')::
+
+    CSP           = 1*(WSP / ("\\" LF))
+
+Versions may be specified according to the PEP-440 [#pep440]_ rules. (Note:
+URI is defined in std-66 [#std66]_::
+
+    version-cmp   = "<" / "<=" / "!=" / "==" / ">=" / ">" / "~=" / "==="
+    version       = 1*( DIGIT / ALPHA / "-" / "_" / "." / "*" )
+    versionspec   = ["("] version-cmp version *(',' version-cmp version) [")"]
+    urlspec       = "@" URI
+
+Environment markers allow making a specification only take effect in some
+environments::
+
+    marker-op     = version-cmp / "in" / "not in"
+    python-str-c  = (WSP / ALPHA / DIGIT / "(" / ")" / "." / "{" / "}" /
+                    "-" / "_" / "*"
+    python-str    = "'" *(python-str-c / DQUOTE) "'"
+    python-str    =/ DQUOTE *(python-str-c / "'") DQUOTE
+    marker-var    = python-str / "python_version" / "python_full_version" /
+                    "os_name"" / "sys_platform" / "platform_release" /
+                    "platform_version" / "platform_machine" /
+                    "platform_python_implementation" / "implementation_name" /
+                    "implementation_version" / "platform_dist_name"
+                    "platform_dist_version" / "platform_dist_id"
+    marker-expr   = "(" marker ")" / (marker-var [marker-op marker-var])
+    marker        = marker-expr *( ("and" / "or") marker-expr)
+    name-marker   = ";" *CSP marker
+    url-marker    = ";" 1*CSP marker
+
+Optional components of a distribution may be specified using the extras
+field::
+
+    identifier    = 1*( DIGIT / ALPHA / "-" / "_" )
+    name          = identifier
+    extras        = "[" identifier *("," identifier) "]"
+
+Giving us a rule for name based requirements::
+
+    name_req      = name [CSP extras] [CSP versionspec] [CSP name-marker]
+
+And a rule for direct reference specifications::
+
+    url_req       = name [CSP extras] urlspec [CSP url-marker]
+
+Leading to the unified rule that can specify a dependency::
+
+    specification = (name_req / location_req) [CSP comment]
+
+Whitespace
+----------
+
+Non line-breaking whitespace is optional and has no semantic meaning.
```

Comment:
Ah, well, that changes things :-).

If all whitespace is optional, then why is `CSP` specifically defined as "at least one whitespace character" instead of "zero or more whitespace characters", and why do `name-marker` and `url-marker` differ only in that the latter has required whitespace?